components/lsp-form/MintForm.tsx

```diff
@@ -9,19 +9,109 @@ import {
 } from "./LSPForm.styled";
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 import { faArrowDown } from "@fortawesome/free-solid-svg-icons";
+import { ethers } from "ethers";
+import toWeiSafe from "../../utils/convertToWeiSafely";
 
 import LongShort from "./LongShort";
 import Collateral from "./Collateral";
+import { RefetchOptions, QueryObserverResult } from "react-query";
+import { TokensCreated } from "./useTokensCreatedEvents";
+// max uint value is 2^256 - 1
+const MAX_UINT_VAL = ethers.constants.MaxUint256;
+const INFINITE_APPROVAL_AMOUNT = MAX_UINT_VAL;
 
 interface Props {
+  address: string;
+  contractAddress: string;
   setShowSettle: React.Dispatch<React.SetStateAction<boolean>>;
+  lspContract: ethers.Contract | null;
+  erc20Contract: ethers.Contract | null;
+  web3Provider: ethers.providers.Web3Provider | null;
+  collateralBalance: ethers.BigNumber;
+  tokensMinted: ethers.BigNumber;
+  collateralPerPair: string;
+  refetchTokensCreatedEvents: (
+    options?: RefetchOptions | undefined
+  ) => Promise<
+    QueryObserverResult<void | TokensCreated[] | undefined, unknown>
+  >;
+  setCollateralBalance: React.Dispatch<React.SetStateAction<ethers.BigNumber>>;
+  erc20Decimals: string;
+  collateralDecimals: string;
 }
-const MintForm: FC<Props> = ({ setShowSettle }) => {
-  const [collateral, setCollateral] = useState("");
+
+const MintForm: FC<Props> = ({
+  // setShowSettle,
+  lspContract,
+  erc20Contract,
+  contractAddress,
+  collateralBalance,
+  tokensMinted,
+  collateralPerPair,
+  refetchTokensCreatedEvents,
+  address,
+  setCollateralBalance,
+  erc20Decimals,
+  collateralDecimals,
+}) => {
+  const [collateral, setCollateral] = useState("uma");
   const [amount, setAmount] = useState("");
   const [longTokenAmount, setLongTokenAmount] = useState("");
   const [shortTokenAmount, setShortTokenAmount] = useState("");
 
+  const mint = useCallback(async () => {
+    if (lspContract && erc20Contract && amount) {
+      const weiAmount = toWeiSafe(amount);
+      try {
+        const allowance = await erc20Contract.allowance(
+          address,
+          contractAddress
+        );
+        const balance = await erc20Contract.balanceOf(address);
+        const hasToApprove = allowance.lt(balance);
+        if (hasToApprove) {
+          const approveTx = await erc20Contract.approve(
+            contractAddress,
+            INFINITE_APPROVAL_AMOUNT
+          );
+
+          await approveTx.wait();
+        }
+        // Need to send the correct amount based on the collateral pair ** the amount
+        // User has specified in the input.
+        // Get rid of float, then divide it out later.
+        const ratio = ethers.BigNumber.from(1 / Number(collateralPerPair)).mul(
```

Comment:
collateralPerPair is a string. It's unclear whether it's in wei or not, but if it is, this conversion can lose precision.

`1 / collateralPerPair` is also likely to lose precision in certain cases. We should do all of this math using a high-precision math library.

My suggestion is to decide up-front if you want all your numbers to be represented as true decimals using some sort of library like bignumber.js (which handles decimals) OR if you want to mirror solidity math and represent decimals as fixed point numbers that are multiplied up.

My suggestion for this line if you want to have these numbers use solidity-like math is the following

```ts
const toBN = ethers.BigNumber.from;
const ratio = toBN(10).pow(18).div(toBN(collateralPerPair));
```

Doing all multiplications first and all divisions last ensures that you lose as little precision as possible.

Note: to ensure you don't end up with rounding issues, you may have to compute the ratio in the create line below. Otherwise, the ratio may get rounded as an intermediate step and you may end up passing a slightly innacurate value to create. Something like the following may work a little better:

```ts
// Note: this assumes that collateralPerPairInWei is the raw value from the contract (i.e. 1e18 * decimal value)
const toBN = ethers.BigNumber.from;
const scaling = toBN(10).pow(18);
lspContract.create(weiAmount.mul(scaling).div(collateralPerPairInWei)).then(...)
```

Reply:
CollateralPerPair is the value returned from the contract. I just converted it to a string. I can keep it as a big number if you'd prefer.

Reply:
It looks like you're converting it to a decimal before setting it, though. So it's the value returned by the contract divided by 1e18, right?

Either way, we need to be careful about precision loss (see above). This math needs to be updated to avoid precision loss or it will be susceptible to rounding errors that cause reverts.